src/utils/map.hpp

```diff
@@ -0,0 +1,963 @@
+    }
+
+    /**
+     * @brief This function finds the value corresponding to key and returns it.
+     *
+     * This function will attempt to find the given key in the map. If it is successful,
+     * it will return the object by reference. If it cannot locate it, it will construct
+     * a new object that corresponds to the given key and inserts both the key and value.
+     *
+     * @return Value that corresponds to the given key by reference.
+     */
+    template <typename key, typename value, typename Hash>
+    value& map<key,value,Hash>::operator[](const key& k)
+    {
+        iterator it = find(k);
+        if(it.isNull())
+        {
+            it = insert(k,value());
+        }
+        return *it;
+    }
+
+    /**
+     * @brief This function finds the value corresponding to a temporary key and returns it.
+     *
+     * This function will attempt to find the given key in the map. If it is successful,
+     * it will return the object by reference. If it cannot locate it, it will construct
+     * a new object that corresponds to the given key and insert them both into the map.
+     *
+     * @return Value that corresponds to the given key by reference.
+     */
+    template <typename key, typename value, typename Hash>
+    value& map<key,value,Hash>::operator[] (key&& k)
+    {
+        iterator it = find(k);
+        if(it.isNull())
+        {
+            it = insert(k,value());
+        }
+        return *it;
+    }
+
+    /**
+     * This function iterates through the map and prints the keys and value
+     * to the ostream. This is implemented with the intent of error logging
+     */
+    template <typename k, typename v, typename Hash>
+    std::ostream& operator<<(std::ostream& out, const map<k, v, Hash>& d)
+    {
+        out << "The key value combos inside hash map are: \n";
+        for(int i = 0; i < d.m_capacity; ++i)
+        {
+            if(d.m_map[i]&& d.m_map[i]->available != true)
+            out << "Hash: "<<std::setw(2)<<std::right
+                << i <<" Key: " <<std::setw(3)<< d.m_map[i]->nodeKey
+                << " Value: " << d.m_map[i]->nodeValue << "\n";
+        }
+        return out;
+    }
+
+    /**
+     * @brief This function insert a key and corresponding value into the map
+     *
+     * This function will attempt to find the given key in the map. If it is successful,
+     * it will replace the old value with the new one. If it cannot locate it, it will construct
+     * a new object that corresponds to the given key and insert them both into the map.
+     *
+     * @return iterator corresponding to the position of the inserted key
+     */
+    template <typename key, typename value, typename Hash>
+    typename map<key,value,Hash>::iterator map<key,value,Hash>::insert(const key& k,const value& v)
+    {
+        int findPosition = -1;
+        node* found = insertFind(k, findPosition);
+        if(found) found->available = false;
```

Comment:
`found->available` can't be true? Why is there no statement for making `found->available` true?